src/lib/merkle_mask/masking_merkle_tree.ml

```diff
@@ -0,0 +1,247 @@
+
+  let create () =
+    { parent= ref None
+    ; account_tbl= Location.Table.create ()
+    ; hash_tbl= Addr.Table.create () }
+
+  (* getter, setter, so we don't rely on a particular implementation *)
+  let find_account t location = Location.Table.find t.account_tbl location
+
+  let set_account t location account =
+    Location.Table.set t.account_tbl ~key:location ~data:account
+
+  let remove_account t location = Location.Table.remove t.account_tbl location
+
+  (* don't rely on a particular implementation *)
+  let find_hash t address = Addr.Table.find t.hash_tbl address
+
+  let set_hash t address hash =
+    Addr.Table.set t.hash_tbl ~key:address ~data:hash
+
+  (* a read does a lookup in the account_tbl; if that fails, delegate to parent *)
+  let get t location =
+    if not (has_parent t) then failwith "get: mask does not have a parent" ;
+    match find_account t location with
+    | Some account -> Some account
+    | None -> Base.get (get_parent t) location
+
+  (* for tests, observe whether location is in mask *)
+  let location_in_mask t location = Option.is_some (find_account t location)
+
+  (* given a Merkle path given by the mask parent and an account address, calculate addresses and hash for each node affected 
+     by the account hash; that is, along the path from the account address to root
+   *)
+  let addresses_and_hashes_from_merkle_path t merkle_path account_address
+      account_hash : (Addr.t * Hash.t) list =
+    let get_addresses_hashes height accum node =
+      let last_address, last_hash =
+        match List.hd accum with
+        | Some elt -> elt
+        | None -> failwith "addresses_and_hashes_from_merkle_path: empty accum"
+      in
+      let next_address =
+        match Addr.parent last_address with
+        | Ok addr -> addr
+        | Error _s ->
+            failwith
+              "addresses_and_hashes_from_merkle_path: could not get next \
+               address"
+      in
+      (* the Merkle path is provided by the mask's parent; some hashes may be out-of-date after
+         the mask is updated. Check whether our hash mask has an entry for corresponding address 
+         in the path, and use it if present
+         *)
+      let merkle_node_address = Addr.sibling last_address in
+      let mask_hash = find_hash t merkle_node_address in
+      match node with
+      | `Left h ->
+          let sibling_hash = Option.value mask_hash ~default:h in
+          (next_address, Hash.merge ~height last_hash sibling_hash) :: accum
+      | `Right h ->
+          let sibling_hash = Option.value mask_hash ~default:h in
+          (next_address, Hash.merge ~height sibling_hash last_hash) :: accum
+    in
+    List.foldi merkle_path
+      ~init:[(account_address, account_hash)]
+      ~f:get_addresses_hashes
+
+  (* a write writes only to the mask, parent is not involved 
+     need to update both account and hash pieces of the mask
+     *)
+  let set t location account =
```

Comment:
Removed the tests which checks for the parent for `set` and `get`, since that's now enforced by the types.

The base interface requires a `create` for `Attached`. Since you can't guess what the parent is in that case, it always fails (with an error message that should be helpful).